Applications/Spire/Source/Ui/DropDownMenu2.cpp

```diff
@@ -0,0 +1,250 @@
+    }();
+    menu.set_current(index);
+  }
+}
+
+DropDownMenu2::DropDownMenu2(std::vector<DropDownItem*> items,
+    QWidget* parent)
+    : QWidget(parent, Qt::Tool | Qt::FramelessWindowHint),
+      m_selected_index(0),
+      m_max_displayed_items(5) {
+  setAttribute(Qt::WA_ShowWithoutActivating);
+  m_shadow = new DropShadow(true, false, this);
+  setObjectName("DropDownMenu");
+  setStyleSheet(R"(
+    #DropDownMenu {
+      background-color: #C8C8C8;
+    })");
+  auto layout = new QVBoxLayout(this);
+  layout->setContentsMargins(scale_width(1), 0, scale_width(1),
+    scale_height(1));
+  m_scroll_area = new ScrollArea(this);
+  m_scroll_area->setFocusProxy(parent);
+  m_scroll_area->setWidgetResizable(true);
+  layout->addWidget(m_scroll_area);
+  auto body = new QWidget(m_scroll_area);
+  m_list_layout = new QVBoxLayout(body);
+  m_list_layout->setContentsMargins({});
+  m_list_layout->setSpacing(0);
+  m_scroll_area->setWidget(body);
+  for(auto& item : items) {
+    m_list_layout->addWidget(item);
+    m_item_hovered_connections.AddConnection(item->connect_highlighted_signal(
+      [=] (const auto& value) { m_hovered_signal(item->get_value()); }));
+    m_item_selected_connections.AddConnection(item->connect_selected_signal(
+      [=] (const auto& value) { on_item_selected(item->get_value(), item); }));
+  }
+  if(m_list_layout->count() > 0) {
+    update_height();
+    static_cast<DropDownItem*>(
+      m_list_layout->itemAt(0)->widget())->set_highlight();
+    m_selected_index = 0;
+    m_current_index = 0;
+    m_selected_signal(get_value(*m_selected_index));
+    m_current_signal(get_value(*m_current_index));
+  }
+  parent->installEventFilter(this);
+  parent->window()->installEventFilter(this);
+  hide();
+}
+
+bool DropDownMenu2::event(QEvent* event) {
+  if(event->type() == QEvent::WindowDeactivate &&
+      focusWidget() != nullptr && !isAncestorOf(focusWidget())) {
+    hide();
+  }
+  return QWidget::event(event);
+}
+
+bool DropDownMenu2::eventFilter(QObject* watched, QEvent* event) {
+  if(watched == parent()) {
+    switch(event->type()) {
+      case QEvent::Move:
+        move_to_parent();
+        break;
+      case QEvent::FocusOut:
+        if(!isActiveWindow()) {
+          hide();
+        }
+        break;
+      case QEvent::Resize:
+        setFixedWidth(parentWidget()->width());
+        break;
+    }
+  } else if(watched == parentWidget()->window()) {
+    if(event->type() == QEvent::WindowDeactivate && !isActiveWindow()) {
+      hide();
+    } else if(event->type() == QEvent::Move) {
+      move_to_parent();
+    } else if(event->type() == QEvent::MouseButtonPress) {
+      auto e = static_cast<QMouseEvent*>(event);
+      if(e->button() == Qt::LeftButton) {
+        hide();
+      }
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+void DropDownMenu2::hideEvent(QHideEvent* event) {
+  if(m_current_index) {
+    get_item(*m_current_index)->reset_highlight();
+    m_current_index = none;
```

Comment:
Why does this happen? If the menu is hidden I see no reason to reset the current item.

Reply:
You mentioned using the CurrentSignal with an invalid variant as the menu closed signal. I'm not sure how else to implement that without having the menu's state be different from what the signal is signaling.

Reply:
I am wondering what purpose this entire method serves. I think this method should be removed entirely.

Reply:
That's fine, but the inputs will need to define some way to signal that the menu was closed.

Reply:
I don't know what "inputs" are. Is an input a ```DropDownMenuItem```?

Reply:
The inputs would be the parent button, text input, etc. Basically, whatever is controlling the DropDownMenu.

Reply:
There is a ```ClosedSignal``` that indicates that the ```DropDownMenu``` is closed.